script.py

```diff
@@ -0,0 +1,186 @@
+
+    exploded_df = (unichem_molecule_struct_spark_df
+                   .select('MOLECULE_PDB_ID',
+                           'MOLECULE_CHEMBL_ID',
+                           f.explode(unichem_molecule_struct_spark_df.STRUCTURE_ID)))
+
+    # print('----- STRUCTURE & TARGET EXPLODED -----')
+    # exploded_df.show()
+
+    chain_ensembl_struct_mol_joined = (pdb_chain_ensembl
+                                       .join(exploded_df, pdb_chain_ensembl["PDB"] == exploded_df["col"])
+                                       .drop('col'))
+
+    return chain_ensembl_struct_mol_joined
+
+
+def main():
+
+    # TIMER 1 START
+    start_1 = timeit.default_timer()
+
+    # ----- GET MOLECULE ID -----
+    # Molecule DataFrame
+    molecule_df = (
+        spark.read
+        .parquet(args.parquet)
+        .select(f.col('id'))
+        .withColumnRenamed('id', 'MOLECULE_CHEMBL_ID')
+    )
+
+    # ----- GET PDB ID -----
+    # Unichem molecules DataFrame
+    unichem_df = spark.read.csv(args.unichem, sep=r'\t', header=True)
+
+    # Join Unichem and Molecule DataFrame
+    unichem_molecule_df = (molecule_df
+                           .join(unichem_df, unichem_df["From src:'1'"] == molecule_df['MOLECULE_CHEMBL_ID'])
+                           .withColumnRenamed("To src:'3'", 'MOLECULE_PDB_ID')
+                           .drop("From src:'1'")
+                           )
+
+    # TIMER 2 START
+    start_2 = timeit.default_timer()
+
+    # ----- GET STRUCTURE ID -----
+    # Apply get_structure function on the Unichem-Molecule DataFrame
+
+    unichem_molecule_struct_pd_df = (unichem_molecule_df
+                                     .toPandas()
+                                     .assign(STRUCTURE_ID=lambda x: x['MOLECULE_PDB_ID'].apply(get_structure))
+                                     )
+
+    unichem_molecule_struct_pd_df.to_csv('structure_of_molecules.csv', index=False, header=True)
+
+    # Convert Spark DataFrame into Pandas DataFrame
+    unichem_molecule_struct_spark_df = spark.createDataFrame(unichem_molecule_struct_pd_df)
+
+    # Count number of structure per molecule
+    unichem_molecule_struct_spark_df = (unichem_molecule_struct_spark_df
+                                        .withColumn("NB_OF_STRUCT_PER_MOL", f.size(f.col("STRUCTURE_ID")))
+                                        .orderBy(f.desc("NB_OF_STRUCT_PER_MOL"))
+                                        )
+
+    unichem_molecule_struct_spark_df = (unichem_molecule_struct_spark_df
+                                        .filter(f.col('MOLECULE_CHEMBL_ID') != 'CHEMBL692')
+                                        .filter(f.col('MOLECULE_CHEMBL_ID') != 'CHEMBL1236970')
+                                        .filter(f.col('MOLECULE_CHEMBL_ID') != 'CHEMBL457299')
+                                        .filter(f.col('MOLECULE_CHEMBL_ID') != 'CHEMBL113178')
+                                        .select('MOLECULE_CHEMBL_ID', 'MOLECULE_PDB_ID', 'STRUCTURE_ID'))
```

Comment:
Multiple filters should be collapsed into a single expression:

```suggestion
unichem_molecule_struct_spark_df = (
    unichem_molecule_struct_spark_df
    .filter(
        (f.col('MOLECULE_CHEMBL_ID') != 'CHEMBL692') & 
        (f.col('MOLECULE_CHEMBL_ID') != 'CHEMBL1236970') &
        (f.col('MOLECULE_CHEMBL_ID') != 'CHEMBL457299') & 
        (f.col('MOLECULE_CHEMBL_ID') != 'CHEMBL113178')
    )
    .select('MOLECULE_CHEMBL_ID', 'MOLECULE_PDB_ID', 'STRUCTURE_ID')
)
```

However we should create a filter array. You can imagine how complicated this experssion would be if there would 20 molecules to exclude.

```python
dropped_molecules = [
    'CHEMBL692', 'CHEMBL1236970', 'CHEMBL457299', 'CHEMBL113178'
]

unichem_molecule_struct_spark_df = (
    unichem_molecule_struct_spark_df
    .filter(~col('MOLECULE_CHEMBL_ID').isIn(dropped_molecules))
    .select('MOLECULE_CHEMBL_ID', 'MOLECULE_PDB_ID', 'STRUCTURE_ID')
)
```

Reply:
`isin` worked for me instead of `isIn`